bittide/src/Bittide/Transceiver.hs

```diff
@@ -7,453 +7,556 @@
+  -- neighbor might decide to not send user data for a long time, even if this
+  -- is asserted.
+  }
+
+data Inputs tx rx ref free rxS n = Inputs
+  { clock :: Clock free
+  -- ^ See 'Input.clock'
+  , reset :: Reset free
+  -- ^ See 'Input.reset'
+  , refClock :: Clock ref
+  -- ^ See 'Input.refClock'
+  , channelNames :: Vec n String
+  -- ^ See 'Input.channel'
+  , clockPaths :: Vec n String
+  -- ^ See 'Input.clockPath'
+  , rxNs :: Signal rxS (BitVector n)
+  -- ^ See 'Input.rxN'
+  , rxPs :: Signal rxS (BitVector n)
+  -- ^ See 'Input.rxP'
+  , txDatas :: Vec n (Signal tx (BitVector 64))
+  -- ^ See 'Input.txData'
+  , txReadys :: Vec n (Signal tx Bool)
+  -- ^ See 'Input.txReady'
+  , rxReadys :: Vec n (Signal rx Bool)
+  -- ^ See 'Input.rxReady'
   }
 
+
 transceiverPrbsN ::
-  forall tx rx refclk freeclk txS rxS chansUsed .
-  ( KnownNat chansUsed
+  forall tx rx ref free txS rxS n .
+  ( KnownNat n
   , HasSynchronousReset tx
   , HasDefinedInitialValues tx
 
   , HasSynchronousReset rx
   , HasDefinedInitialValues rx
 
-  , HasSynchronousReset freeclk
-  , HasDefinedInitialValues freeclk
-  ) =>
-  Clock refclk ->
-  Clock freeclk ->
-
-  Reset freeclk ->
+  , HasSynchronousReset free
+  , HasDefinedInitialValues free
 
-  Vec chansUsed String ->
-  Vec chansUsed String ->
+  , KnownDomain rxS
+  , KnownDomain txS
+  , KnownDomain ref
+  , KnownDomain free
+  ) =>
+  Config free ->
+  Inputs tx rx ref free rxS n ->
+  Outputs n tx rx txS free
+transceiverPrbsN opts inputs@Inputs{clock, reset, refClock} = Outputs
+  -- tx
+  { txClocks          = map (.txClock) outputs
+  , txResets          = map (.txReset) outputs
+  , txReadys          = map (.txReady) outputs
+  , txSamplings       = map (.txSampling) outputs
+
+  -- rx
+  , rxClocks = map (.rxClock) outputs
+  , rxResets = map (.rxReset) outputs
+  , rxDatas  = map (.rxData) outputs
 
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
+  -- transceiver
+  , txPs = pack <$> bundle (map (.txP) outputs)
+  , txNs = pack <$> bundle (map (.txN) outputs)
 
-  TransceiverOutputs chansUsed tx rx txS freeclk
-transceiverPrbsN refclk freeclk rst chanNms clkPaths rxns rxps = TransceiverOutputs
-  { txClocks = map (.txClock) outputs
-  , rxClocks = map (.rxClock) outputs
-  , txPs     = map (.txP)     outputs
-  , txNs     = map (.txN)     outputs
-  , linkUps  = map (.linkUp)  outputs
-  , stats    = map (.stats)   outputs
+  -- free
+  , linkUps    = map (.linkUp)  outputs
+  , linkReadys = map (.linkReady)  outputs
+  , stats      = map (.stats)   outputs
   }
  where
-  outputs = zipWith4 (transceiverPrbs refclk freeclk rst) chanNms clkPaths rxns rxps
+  -- XXX: Replacing 'zipWithN' with '<$>' and '<*>' triggers a combination of:
+  --
+  --       * https://github.com/clash-lang/clash-compiler/issues/2723
+  --       * https://github.com/clash-lang/clash-compiler/issues/2722
+  --
+  outputs = zipWith8 go
+    (iterateI (+1) 0)
```

Comment:
Almost! That's actually what I tried to do initially, but

```
indicesI :: KnownNat n => Vec n (Index n)
```

..so it will require the number of FPGAs instantiated to be the same as the debug Index. I.e., that will mean extra constraints and conversion functions which seems a bit much for a debugging feature.

Reply:
That's fair. I didn't consider that point. 

Nevertheless, we should add a note somewhere then, as I am a bit worried that we simply oversee the `Unsigned 3` once we change the setup to more than 8 nodes and run into issues then.

Reply:
I can also just removed the feature entirely. It's just for debugging / sanity checking, the actual code does not rely on it.

Reply:
No. I would keep it, as you only introduce extra work in case you need it again in the feature otherwise. Just add a note to the comment.